### PR TITLE
[SPARK-8021] [SQL] [PYSPARK] make Python read/write API consistent with Scala

### PR DESCRIPTION
**Pull request by davies**
add schema()/format()/options() for reader,  add mode()/format()/options()/partitionBy() for writer

cc @rxin @yhuai  @pwendell

### CODE REVIEW

**Comment by rxin** [comment]
lol I've been working on this too in the last 20 mins.

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -194,6 +219,48 @@ def __init__(self, df):
         self._sqlContext = df.sql_ctx
         self._jwrite = df._jdf.write()
 
+    def mode(self, saveMode):
```
Comment (rxin) [inline python/pyspark/sql/readwriter.py]: since 1.4?

File: python/pyspark/sql/readwriter.py [hunk 2]
```diff
@@ -194,6 +219,48 @@ def __init__(self, df):
...
+    @since(1.4)
+    def partitionBy(self, *cols):
+        """
+        Partitions the output by the given columns on the file system.
+        If specified, the output is laid out on the file system similar
+        to Hive's partitioning scheme.
+
+        :param cols: name of columns
+        """
+        self._jwrite = self._jwrite.partitionBy(_to_seq(self._sqlContext._sc, cols))
```
Comment (rxin) [inline python/pyspark/sql/readwriter.py]: would be great to support the 1st argument being a list

**Comment by rxin** [comment]
LGTM. We should add tests too (maybe later so we can make this in 1.4)

**Comment by pwendell** [comment]
Jenkins, retest this please.

**Comment by pwendell** [comment]
This failed a totally unrelated streaming test, so I think it's okay.

**Comment by rxin** [comment]
@pwendell but Python hasn't even run yet because it runs after the Scala streaming tests

**Pull request closed without merging**